# Proyek Machine Learning Terapan Terakhir - Lalu Habib Satya Wiguna

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import warnings 

warnings.filterwarnings('ignore')
%matplotlib inline

## Domain Knowledge
Langkah pertama yang kita lakukan adalah melihat isi dataset yang akan digunakan serta memahami setiap atribut yang ada pada dataset tersebut. Dataset yang digunakan yaitu [Goodreads-books](https://www.kaggle.com/datasets/jealousleopard/goodreadsbooks) yang didapaatkan dari situs [kaggle](https://www.kaggle.com)

In [2]:
books = pd.read_csv('dataset/books.csv', error_bad_lines=False, warn_bad_lines=False)
books

bookID                                              title  \
0           1  Harry Potter and the Half-Blood Prince (Harry ...   
1           2  Harry Potter and the Order of the Phoenix (Har...   
2           4  Harry Potter and the Chamber of Secrets (Harry...   
3           5  Harry Potter and the Prisoner of Azkaban (Harr...   
4           8  Harry Potter Boxed Set  Books 1-5 (Harry Potte...   
...       ...                                                ...   
11118   45631   Expelled from Eden: A William T. Vollmann Reader   
11119   45633                        You Bright and Risen Angels   
11120   45634                    The Ice-Shirt (Seven Dreams #1)   
11121   45639                                        Poor People   
11122   45641                        Las aventuras de Tom Sawyer   

                                                 authors  average_rating  \
0                             J.K. Rowling/Mary GrandPré            4.57   
1                             J.K. Rowling/Mary GrandPré            4.49   
2                                           J.K. Rowling            4.42   
3                             J.K. Rowling/Mary GrandPré            4.56   
4                             J.K. Rowling/Mary GrandPré            4.78   
...                                                  ...             ...   
11118  William T. Vollmann/Larry McCaffery/Michael He...            4.06   
11119                                William T. Vollmann            4.08   
11120                                William T. Vollmann            3.96   
11121                                William T. Vollmann            3.72   
11122                                         Mark Twain            3.91   

             isbn         isbn13 language_code    num_pages  ratings_count  \
0      0439785960  9780439785969           eng          652        2095690   
1      0439358078  9780439358071           eng          870        2153167   
2      0439554896  9780439554893           eng          352           6333   
3      043965548X  9780439655484           eng          435        2339585   
4      0439682584  9780439682589           eng         2690          41428   
...           ...            ...           ...          ...            ...   
11118  1560254416  9781560254416           eng          512            156   
11119  0140110879  9780140110876           eng          635            783   
11120  0140131965  9780140131963           eng          415            820   
11121  0060878827  9780060878825           eng          434            769   
11122  8497646983  9788497646987           spa          272            113   

       text_reviews_count publication_date        publisher  
0                   27591        9/16/2006  Scholastic Inc.  
1                   29221         9/1/2004  Scholastic Inc.  
2                     244        11/1/2003       Scholastic  
3                   36325         5/1/2004  Scholastic Inc.  
4                     164        9/13/2004       Scholastic  
...                   ...              ...              ...  
11118                  20       12/21/2004    Da Capo Press  
11119                  56        12/1/1988    Penguin Books  
11120                  95         8/1/1993    Penguin Books  
11121                 139        2/27/2007             Ecco  
11122                  12        5/28/2006    Edimat Libros  

[11123 rows x 12 columns]

Dari output di atas kita mempunyai 11123 data dengan 12 kolom pada Goodreads-books dataset. Berdasarkan informasi dari Kaggle, variabel-variabel pada Goodreads-books dataset adalah sebagai berikut:  
- bookID : kode unik dari setiap buku.
- title : nama atau judul dari buku.
- authors : nama dari penulis buku.
- average_rating : rata-rata penilaian dari setiap buku.
- isbn : kode unik buku yang merujuk pada ISBN (International Standard Book Number).
- isbn13 : 13-digit ISBN untuk mengidentifikasi buku.
- language_code : kode bahasa dari buku, misalnya 'eng' untuk English.
- num_pages : jumlah halaman dari buku.
- ratings_count : total penilaian yang didapatkan dari suatu buku.
- text_reviews_count : total teks review yang diperoleh oleh suatu buku.
- publication_date : tanggal buku dirilis ke publik.
- publisher : nama penerbit yang menerbitkan buku.

In [3]:
books.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 11123 entries, 0 to 11122
Data columns (total 12 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   bookID              11123 non-null  int64  
 1   title               11123 non-null  object 
 2   authors             11123 non-null  object 
 3   average_rating      11123 non-null  float64
 4   isbn                11123 non-null  object 
 5   isbn13              11123 non-null  int64  
 6   language_code       11123 non-null  object 
 7     num_pages         11123 non-null  int64  
 8   ratings_count       11123 non-null  int64  
 9   text_reviews_count  11123 non-null  int64  
 10  publication_date    11123 non-null  object 
 11  publisher           11123 non-null  object 
dtypes: float64(1), int64(5), object(6)
memory usage: 1.0+ MB


Dari output di atas dapat dilihat bahwa:

- Terdapat 6 kolom dengan tipe object, yaitu: title, authors, isbn, language_code, publication_date, publisher.
- Terdapat 5 kolom numerik dengan tipe int64, yaitu: bookID, isbn13, num_pages, ratings_count, text_review_count.
- Terdapat 1 kolom numerik dengan tipe float64, yaitu: average_rating.

In [4]:
list(books)

['bookID',
 'title',
 'authors',
 'average_rating',
 'isbn',
 'isbn13',
 'language_code',
 '  num_pages',
 'ratings_count',
 'text_reviews_count',
 'publication_date',
 'publisher']

Dapat dilihat pada output di atas bahwa ada variabel yang penulisannya memiliki spasi di awal yaitu num_pages. Oleh karena itu kita perlu memperbaikinya agar lebih mudah untuk diakses nanti.

In [5]:
# menghapus spasi pada variabel num_pages agar memudahkan pengaksesan variabel nanti

books.columns = books.columns.str.strip()
list(books)

['bookID',
 'title',
 'authors',
 'average_rating',
 'isbn',
 'isbn13',
 'language_code',
 'num_pages',
 'ratings_count',
 'text_reviews_count',
 'publication_date',
 'publisher']

In [6]:
books.describe()

bookID  average_rating        isbn13     num_pages  \
count  11123.000000    11123.000000  1.112300e+04  11123.000000   
mean   21310.856963        3.934075  9.759880e+12    336.405556   
std    13094.727252        0.350485  4.429758e+11    241.152626   
min        1.000000        0.000000  8.987060e+09      0.000000   
25%    10277.500000        3.770000  9.780345e+12    192.000000   
50%    20287.000000        3.960000  9.780582e+12    299.000000   
75%    32104.500000        4.140000  9.780872e+12    416.000000   
max    45641.000000        5.000000  9.790008e+12   6576.000000   

       ratings_count  text_reviews_count  
count   1.112300e+04        11123.000000  
mean    1.794285e+04          542.048099  
std     1.124992e+05         2576.619589  
min     0.000000e+00            0.000000  
25%     1.040000e+02            9.000000  
50%     7.450000e+02           47.000000  
75%     5.000500e+03          238.000000  
max     4.597666e+06        94265.000000

Fungsi describe() memberikan informasi statistik pada masing-masing kolom, antara lain:

- Count adalah jumlah sampel pada data.
- Mean adalah nilai rata-rata.
- Std adalah standar deviasi.
- Min yaitu nilai minimum setiap kolom.
- 25% adalah kuartil pertama. Kuartil adalah nilai yang menandai batas interval - dalam empat bagian sebaran yang sama.
- 50% adalah kuartil kedua, atau biasa juga disebut median (nilai tengah).
- 75% adalah kuartil ketiga.
- Max adalah nilai maksimum.

## Data Cleaning
Pada tahap ini, kita akan melakukan pembersihan pada data yaitu dengan menghapus data duplikat (jika ada), melihat data yang bernilai kosong, serta memperbaiki isi pada beberapa kolom agar lebih mudah membaca datanya.

### Checking Null Value
Pada tahap ini kita perlu memastikan apakah ada data dengan nilai kosong pada dataset kita karena akan memiliki pengaruh pada proses pembuatan model rekomendasi nanti jika data kita memiliki informasi yang kurang.

In [7]:
# melihat apakah ada data yang memiliki nilai kosong

books.isnull().sum()

bookID                0
title                 0
authors               0
average_rating        0
isbn                  0
isbn13                0
language_code         0
num_pages             0
ratings_count         0
text_reviews_count    0
publication_date      0
publisher             0
dtype: int64

Dari output di atas kita dapat melihat bahwa seluruh kolom pada dataset kita tidak memiliki nilai kosong atau null value.

### Checking Duplicated Data
Pada tahap ini kita akan memeriksa apakah ada data duplikat pada dataset yang kita gunakan. Jika ada maka kita perlu untuk menghapusnya agar tidak boros informasi serta mengurangi kemungkinan overfitting.

In [8]:
books.duplicated().sum()

0

Tidak ada data duplikat di dalam dataset.

### Fixing the Data
Pada tahap ini kita perlu mengubah isi pada kolom language_code agar memudahkan kita untuk mengetahui bahasa yang digunakan pada suatu buku. Kemudian kita akan menambahkan satu kolom baru di dataset kita yaitu publication_year yang berisi tentang informasi terkait tahun terbit dari suatu buku.

In [9]:
books.language_code.unique()

array(['eng', 'en-US', 'fre', 'spa', 'en-GB', 'mul', 'grc', 'enm',
       'en-CA', 'ger', 'jpn', 'ara', 'nl', 'zho', 'lat', 'por', 'srp',
       'ita', 'rus', 'msa', 'glg', 'wel', 'swe', 'nor', 'tur', 'gla',
       'ale'], dtype=object)

Dari output di atas dapat dilihat bahwa akan sulit untuk mengetahui bahasa dari suatu buku apabila bentuk datanya seperti di atas. Oleh karena itu kita akan mengubahnya agar menjadi lebih mudah dibaca.

In [10]:
books.language_code.replace(
    to_replace=[
        "eng",
        "en-US",
        "fre",
        "spa",
        "en-GB",
        "mul",
        "grc",
        "enm",
        "en-CA",
        "ger",
        "jpn",
        "ara",
        "nl",
        "zho",
        "lat",
        "por",
        "srp",
        "ita",
        "rus",
        "msa",
        "glg",
        "wel",
        "swe",
        "nor",
        "tur",
        "gla",
        "ale",
    ],
    value=[
        "English",
        "US-English",
        "French",
        "Spanish",
        "British-English",
        "Multiple language",
        "Greek",
        "Middle English",
        "Canada-English",
        "German",
        "Japanese",
        "Arabic",
        "Dutch",
        "Chinese",
        "Latvian",
        "Portuguese",
        "Serbian",
        "Initial Teaching language",
        "Russian",
        "Modern Standard Arabic",
        "Galician",
        "Welsh",
        "Swedish",
        "Murik",
        "Turkish",
        "Gaelic",
        "Afro-Asiatic",
    ],
    inplace=True,
)
books.language_code.unique()

array(['English', 'US-English', 'French', 'Spanish', 'British-English',
       'Multiple language', 'Greek', 'Middle English', 'Canada-English',
       'German', 'Japanese', 'Arabic', 'Dutch', 'Chinese', 'Latvian',
       'Portuguese', 'Serbian', 'Initial Teaching language', 'Russian',
       'Modern Standard Arabic', 'Galician', 'Welsh', 'Swedish', 'Murik',
       'Turkish', 'Gaelic', 'Afro-Asiatic'], dtype=object)

Data pada kolom language_code sekarang menjadi lebih mudah dibaca. Selanjutnya kita akan mengubah nama kolom language_code menjadi language saja karena sudah bukan berbentuk kode lagi.

In [12]:
books = books.rename(columns={'language_code':'language'})

Selanjutnya kita akan menambahkan kolom dengan nama publication_year yang berisi tentang informasi terkait tahun terbit dari suatu buku.

In [14]:
books['publication_year'] = books['publication_date'].apply(lambda x: (int)(str(x[-4:])))

In [15]:
books.head()

bookID                                              title  \
0       1  Harry Potter and the Half-Blood Prince (Harry ...   
1       2  Harry Potter and the Order of the Phoenix (Har...   
2       4  Harry Potter and the Chamber of Secrets (Harry...   
3       5  Harry Potter and the Prisoner of Azkaban (Harr...   
4       8  Harry Potter Boxed Set  Books 1-5 (Harry Potte...   

                      authors  average_rating        isbn         isbn13  \
0  J.K. Rowling/Mary GrandPré            4.57  0439785960  9780439785969   
1  J.K. Rowling/Mary GrandPré            4.49  0439358078  9780439358071   
2                J.K. Rowling            4.42  0439554896  9780439554893   
3  J.K. Rowling/Mary GrandPré            4.56  043965548X  9780439655484   
4  J.K. Rowling/Mary GrandPré            4.78  0439682584  9780439682589   

  language  num_pages  ratings_count  text_reviews_count publication_date  \
0  English        652        2095690               27591        9/16/2006   
1  English        870        2153167               29221         9/1/2004   
2  English        352           6333                 244        11/1/2003   
3  English        435        2339585               36325         5/1/2004   
4  English       2690          41428                 164        9/13/2004   

         publisher  publication_year  
0  Scholastic Inc.              2006  
1  Scholastic Inc.              2004  
2       Scholastic              2003  
3  Scholastic Inc.              2004  
4       Scholastic              2004

Dapat dilihat dari output di atas, dataset sekarang menjadi lebih enak untuk dilihat dan lebih mudah dibaca.

## Exploratory Data Analysis
Pada tahap ini, kita akan melakukan ekplorasi terhadap variabel-variabel yang ada pada dataset yang kita gunakan. Pemahaman terhadap variabel ini penting karena akan membantu kita dalam menentukan pendekatan atau algoritma yang cocok untuk data kita.

In [16]:
list(books)

['bookID',
 'title',
 'authors',
 'average_rating',
 'isbn',
 'isbn13',
 'language',
 'num_pages',
 'ratings_count',
 'text_reviews_count',
 'publication_date',
 'publisher',
 'publication_year']